pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go

```diff
@@ -134,10 +125,6 @@ func (w *Watcher) traversePluginDir(dir string) error {
 		}
 
 		switch mode := info.Mode(); {
-		case mode.IsDir():
```

Comment:
dropping this means we won't add subdirectories while walking the tree, right?

Reply:
yes, do we need to support plugin sockets in the subdirectories?

Reply:
I'll defer to the storage/node folks, but I believe we do, yes

Reply:
I would like to hear from storage/node folks, but at least according to https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-management/device-plugin.md#unix-socket, I am under the impression that device plugin server should only create one socket at the socket dir.

Reply:
this watcher is the single watcher instance for all plugins, right? so the path being added in Start is the root plugins folder, which then contains subfolders for individual plugins?

Reply:
That might be a use case and I am not sure whether there are uses depending on this feature, but at least regarding to the device plugins I have seen, such as nvidia device plugin and kubevirt, they all create the socket at the root plugins dir:
* nvidia: https://github.com/NVIDIA/k8s-device-plugin/blob/055e87d0959e1e905c7332c80bb049692e8c11ae/mig-strategy.go#L90
* kubevirt: https://github.com/kubevirt/device-plugin-manager/blob/c9ca4e84bfedb232bf8c9956f6b8594f42c5cee3/pkg/dpm/plugin.go#L54

Reply:
All csi drivers I know of are creating the socket in the top directory, but I don't think this was was explicitly set in stone. If we want to change this behavior, we probably need to put out a deprecation notice and wait for the deprecation period.

cc @saad-ali

Reply:
If we want to retain the current behavior, is it possible to have the reconciler also periodically traverse the registration directory, instead of only operating on fsnotify events?

Reply:
> All csi drivers I know of are creating the socket in the top directory, but I don't think this was was explicitly set in stone. If we want to change this behavior, we probably need to put out a deprecation notice and wait for the deprecation period.
>
> cc @saad-ali

Correct, the assumption was you can create a socket anywhere under the plugin directory (could be under a sub-directory). We should try to retain that behavior.

Reply:
I came up with a question if we really want to support plugin socket in subdirectories.

Let's say `fsWatcher` is already watching `/var/lib/kubelet/device-plugins/`, and then user creates a socket at `/var/lib/kubelet/device-plugins/foo/plugin.sock`. Now please note that `/var/lib/kubelet/device-plugins/foo/` is unknown to `fsWatcher`, so kubelet would not know the `/var/lib/kubelet/device-plugins/foo/plugin.sock` is created.

What should we do to deal with this senario?
1. periodically traverse the registration directory?
2. automatically watch subdirectories when they are created(which may race, i.e. if the socket is created before we watch the subdirectory we would miss the CREATE event)

Reply:
> Now please note that /var/lib/kubelet/device-plugins/foo/ is unknown to fsWatcher

why is it unknown? if `foo` is present at instantiation time, the code in master adds a watch on the `foo` dir in `traversePluginDir`, and if `foo` is created after instantiation time, the watcher traverses it and starts watching in `handleCreateEvent`

> What should we do to deal with this senario?
> ...
> * automatically watch subdirectories when they are created

today, we automatically watch subdirectories when they are created

Reply:
Ahh indeed! I have overlooked the related code, sorry about that.